File: week06/이채은/PGS_가사 검색.java
import java.util.*;

class Solution {
    public int[] solution(String[] words, String[] queries) {
        Map<Integer, List<String>> originWords = new HashMap<>();
        Map<Integer, List<String>> reverseWords = new HashMap<>();

        for(String word : words){
            int length = word.length();

            originWords.computeIfAbsent(length, i -> new ArrayList<>()).add(word);
            reverseWords.computeIfAbsent(length, i -> new ArrayList<>()).add(reverse(word));
        }


        for(int key : originWords.keySet()){
            List<String> list;

            list = originWords.get(key);
            Collections.sort(list);
            list = reverseWords.get(key);
            Collections.sort(list);
        }

        int qLength = queries.length;
        int[] answer = new int[qLength];
        for(int i = 0; i < qLength; i++){
            String query = queries[i];

            // 길이에 맞는 word list 꺼내기
            List<String> list;
            if(query.charAt(0) == '?'){
                list = reverseWords.get(query.length());
                query = reverse(query);
            } else{
                list = originWords.get(query.length());
            }

            // 구간 찾기
            if(list == null) answer[i] = 0;
            else{
                int low = binarySearch(list, query.replace('?',Character.MIN_VALUE));
                int high = binarySearch(list, query.replace('?',Character.MAX_VALUE));
                answer[i] = high - low;
            }
        }

        return answer;
    }

    int binarySearch(List<String> list, String target){
        int left = 0;
        int right = list.size()-1;

        while(left <= right){
            int mid = (left + right) / 2;

            if(list.get(mid).compareTo(target) < 0) left = mid+1;
            else right = mid-1;
        }
        return left;

    }

    String reverse(String str){
        return new StringBuilder(str).reverse().toString();
    }
}

File: week06/한정민/PGS_가사 검색.py
from bisect import bisect_left, bisect_right

def count_by_range(array, left_value, right_value):
    left_index = bisect_left(array, left_value)
    right_index = bisect_right(array, right_value)
    return right_index - left_index

array = [[] for _ in range(10001)]
reversed_array = [[] for _ in range(10001)]

def solution(words, queries):
    answer = []
    for word in words:
        array[len(word)].append(word)
        reversed_array[len(word)].append(word[::-1])
        
    for i in range(10001):
        array[i].sort()
        reversed_array[i].sort()
    
    for query in queries:
        # 접미사에 와일드카드가 있는 경우
        if query[-1] == '?':
            cnt = count_by_range(array[len(query)], query.replace('?', 'a'), query.replace('?', 'z'))
        # 접두두사에 와일드카드가 있는 경우
        else:
            cnt = count_by_range(reversed_array[len(query)], query[::-1].replace('?', 'a'), query[::-1].replace('?', 'z'))
        answer.append(cnt)
        
    return answer